### PR TITLE
feat(Tag): support click handler for filter tag close icon

### LINKED ISSUE #5419: UX bug: Filterable tag should have separate click behavior for tag body and "x"
## Filterable tag should have separate click behavior for tag body and "x"

Carbon's filterable tag currently seems to use a single click behavior for the entire tag, which is problematic because in some use cases we want to be able to click on the tag body separately.

**Actual:** 

On the new data table filter display bar, clicking on the tag body (as opposed to the x) removes the tag. But this is not the experience a user would expect:

Tag click bug

In the future, we may want users to be able to click on a filterable tag to see all other items with that tag - and that is a use case that Watson Knowledge currently has.

Expected:

Only clicking on the "x" click target would result in tag being removed.

Consumer of component can decide what action if any results when click on tag body.

### PR DESCRIPTION
**Pull request by emyarod**
Closes #5419

This PR adds support for a click handler on the filter tag close icon

#### Changelog

**New**

- `onClose` prop support for filter tag

#### Testing / Reviewing

Ensure that the proper events are fired depending on the click target in filter tags

### CODE REVIEW

**Comment by emyarod** [comment]
now that there are 2 distinct click targets for filter tags, should there be an extra tab stop? if so, should there be a visual difference between the tag body on focus and the tag close icon on focus? cc @aagonzales

**Review by aagonzales — COMMENTED** [review]
It looks like clicking on tag area is still activating the x.
Feb-27-2020 16-31-47

**Comment by aagonzales** [comment]
And I don't think the x should add a tab stop, but it should be trigger on keyboard nav by clicking `esc` and/or allow `arrow` + `enter`. I'm not sure what would be the best practice.

**Comment by emyarod** [comment]
it only appears that the x is focused, but it does not fire the same event:

actions

so I'm wondering if there should be a visual difference between tag body focus and close icon focus. but since you said there shouldn't be an added tab stop the current focus behavior of outlining the close icon may be confusing

**Review by aagonzales — CHANGES_REQUESTED** [review]
1. Filter tag color knob isn't working
Feb-28-2020 14-19-20

2. Different focus states for tag area and close `x`
image

3. I would not think the x would be in the tab order, I think it would get excessive to read tab label then read the close function for every tab if there are many. I will defer to the accessibility team on this however. @dakahn any thoughts?

**Comment by emyarod** [comment]
FYI for filter tag color knob https://github.com/carbon-design-system/carbon/issues/3318

I think there would need to be an extra tab stop if we will differentiate tag body actions from close icon actions, otherwise the full component functionality wouldn't be available to keyboard only users (unless pressing <kbd>Esc</kbd> to close while the tag body is focused is sufficient), will await updated guidance

**Comment by dakahn** [comment]
Agreed @emyarod any mouse functionality should have a keyboard only alternative as well

**Comment by abbeyhrt** [comment]
Is there any reason that the tag area needs focus? If it isn't going to dismiss the tag then the only thing that needs to receive focus accessibility-wise would be the close icon, I believe.

**Comment by aagonzales** [comment]
No, I don't think the content needs a focus I just don't want the focus for the x to appear when the user clicks on the content area. The user must click on the x to close the icon and it seems confusing that the focus would appear when not clicking directly on the x.

This is the behavior that needs to be corrected:
Mar-03-2020 11-48-08

**Comment by emyarod** [comment]
ok so in that case we are just moving the only clickable area from the tag body to the close icon?

**Comment by aagonzales** [comment]
Yes 

image

**Comment by emyarod** [comment]
ok since the tag body is currently a `button` element, by default it will have a tab stop. in that case I will change the base element of the tag body to something else. this may be a breaking change if users have been attaching click handlers to the tag body so we will have to announce this change

**Comment by asudoh** [comment]
@emyarod Is this ready for re-review...? Thanks!

**Comment by emyarod** [comment]
@asudoh yes this is ready for re-review

**Review by aagonzales — APPROVED** [review]
Looks good now!

File: packages/components/src/components/tag/_tag.scss [hunk 1]
```diff
@@ -108,28 +106,34 @@
...
+    cursor: pointer;
 
     &:hover {
       background-color: $inverse-hover-ui;
     }
   }
 
-  .#{$prefix}--tag--filter:focus > svg {
+  .#{$prefix}--tag__close-icon svg {
+    fill: $inverse-01;
```
Comment (abbeyhrt) [inline packages/components/src/components/tag/_tag.scss:126]: For my own edification, why did you pull this out? I see it done a lot throughout our styles and I just wanted to learn why, thanks! 😄

**Review by abbeyhrt — APPROVED** [review]
Looks good to me! 🎉

File: packages/components/src/components/tag/_tag.scss (hunk 1 above)
Comment (emyarod) [reply packages/components/src/components/tag/_tag.scss:126]: are you referring to the class (`.bx--tag--filter` to `.bx--tag__close-icon`)? or the `>` selector?

File: packages/react/src/components/Tag/Tag.js [hunk 2]
```diff
@@ -53,8 +58,10 @@ const Tag = ({
       <span className={`${prefix}--tag__label`}>
         {children !== null && children !== undefined ? children : TYPES[type]}
       </span>
-      <Close16 />
-    </button>
+      <button className={`${prefix}--tag__close-icon`} onClick={handleClose}>
```
Comment (joshblack) [inline packages/react/src/components/Tag/Tag.js]: With the button shifting over will we also need to move the `aria-label` over?

Comment (emyarod) [reply packages/react/src/components/Tag/Tag.js]: not sure but the aria label was removed from the close button to the tag body in #4863

**Comment by asudoh** [comment]
@joshblack Any further comments...? Thanks!

**Review by dakahn — CHANGES_REQUESTED** [review]
Tested with JAWS 2020 on Windows 10. The button is read as "clear filter this is not a tag unlabeled 1 button"

Looks like the button under the hood has no label
2020-03-05 17_29_53-Storybook

Worth noting as well is NVDA reads this as "main landmark clickable button" -- so not even picking up the tag text at all there.

And we've also got the expected subsequent DAP error
2020-03-05 17_59_38-Storybook


Adding an `aria-labelledby` attribute referencing an `id` on that wrapper div with the `aria-label` seems to fix both JAWS and NVDA and clear the DAP error though 🎉 
2020-03-05 18_00_43-Storybook

**Review by dakahn — APPROVED** [review]
Sounds good and reads well 👍 DAP errors cleared

**Review by tw15egan — APPROVED** [review]

**Pull request merged**